### PR TITLE
Rate control takes frame complexity into account for better predictions

### PR DESCRIPTION
**Pull request by ateme-developers**
Rate control model updates are now thread safe.
Predictions model is function of the estimated frame complexity.
Prediction model updates itself over time.

### CODE REVIEW

File: Source/Lib/Codec/RateControlModel.c [hunk 1]
```diff
@@ -128,16 +197,55 @@ uint8_t    rate_control_get_quantizer(EbRateControlModel *model_ptr, PicturePare
...
+uint32_t get_inter_qp_for_size(EbRateControlModel *model_ptr, uint32_t desired_size, uint32_t complexity) {
+    uint8_t     qp = 63;
     
-    for (qp = 0; qp < MAX_QP_VALUE; qp++) {
-        float    size = model_ptr->inter_size_predictions[qp];
+    if (!complexity) {
+        for (qp = 0; qp < MAX_QP_VALUE; qp++) {
+            float    size = model_ptr->inter_size_predictions[qp];
+
+            size = (size / (1920 * 1080)) * model_ptr->pixels;
```
Comment (joelsole) [inline Source/Lib/Codec/RateControlModel.c]: is the intention to hard-code these numbers?

File: Source/Lib/Codec/RateControlModel.c [hunk 2]
```diff
@@ -148,49 +256,54 @@ static void record_new_gop(EbRateControlModel *model_ptr, PictureParentControlSe
...
     gop->desired_size = get_gop_size_in_bytes(model_ptr);
-    gop->model_variation = model_ptr->model_variation;
 
-    uint32_t                size = gop->desired_size / model_ptr->intra_period;
+    uint32_t    desired_total_bytes = (model_ptr->desired_bitrate / model_ptr->frame_rate) * model_ptr->reported_frames;
+    int64_t     delta_bytes = desired_total_bytes - model_ptr->total_bytes;
+    int64_t     extra = delta_bytes / 4;
+
+    if (extra > 0 && (uint64_t)extra > (gop->desired_size / 2)) {
+      extra = gop->desired_size / 2;
```
Comment (joelsole) [inline Source/Lib/Codec/RateControlModel.c]: tabs are 4 spaces

File: Source/Lib/Codec/RateControlModel.c (hunk 1 above)
Comment (dynamiccast) [reply Source/Lib/Codec/RateControlModel.c]: The reference model is hard coded here for full HD resolutions: https://github.com/OpenVisualCloud/SVT-AV1/blob/master/Source/Lib/Codec/RateControlModel.c#L33
We proportionate the model's value to the current frame's pixel area. I guess we could use a macro ?

Comment (ateme-developers) [reply Source/Lib/Codec/RateControlModel.c]: The reference model is hard coded here for full HD resolutions: https://github.com/OpenVisualCloud/SVT-AV1/blob/master/Source/Lib/Codec/RateControlModel.c#L33
We proportionate the model's value to the current frame's pixel area. I guess we could use a macro ?

Comment (joelsole) [reply Source/Lib/Codec/RateControlModel.c]: probably better to use macro.

**Comment by hassount** [comment]
@dynamiccast 
Can you please fix the build errors ?

**Comment by hassount** [comment]
@dynamiccast 
Thank you very much :) can you please fix the linux warnings that show up on Travis:
https://travis-ci.com/OpenVisualCloud/SVT-AV1/jobs/180738351

**Comment by dynamiccast** [comment]
@hassount 24eac73 Will fix the issue.

**Comment by hassount** [comment]
@dynamiccast 
Thank you for fixing the build, Now I am getting a crash on linux Ubuntu 18.04 GCC 7.3.0
./SvtAv1EncApp -i /mnt/d/stream/BQSquare_416x240_60.yuv -w 416 -h 240 -fps 60 -rc 1 -tbr 1000000 -n 600

**Comment by ateme-developers** [comment]
Closed in favor of rebase https://github.com/OpenVisualCloud/SVT-AV1/pull/168

**Pull request closed without merging**